### PR TITLE
Add new options to enable storing all classes in SCC

### LINKED ISSUE #19323: Storing all classes to SCC for startup performance
Currently OpenJ9 stores only bootstrap classes and hidden(anon) classes in the SCC by default (i.e` -Xshareclasses` is not specified). As the hidden classes are not from the class path, they are returned by the SCC to the JVM via byte-to-byte comparison, so that we could still benefit from AOT on these classes.

When `-Xshareclasses` is specified, currently only class loader which implements the OpenJ9's public SCC [APIs](https://github.com/eclipse-openj9/openj9/tree/master/jcl/src/openj9.sharedclasses/share/classes/com/ibm/oti/shared) (and its child class loaders) can store classes to SCC. For custom class loaders that do not implement SCC APIs, SCC won't have the class path information so that we don't know where the class is from. These class are not stored to SCC now. 

Similar to hidden(anon) classes, we can store these classes as orphan in the SCC and return via byte-to-byte comparison. If the benefit of AOT outweighs the cost of byte-to-byte comparison, we could still get startup performance improvements.

### PR DESCRIPTION
**Pull request by hangshao0**
1. Add new options -XX:[+/-]ShareOrphans to enable/disable storing classes as orphan to SCC from SCC-unaware class loader(loader does not call SCC helper APIs). This feature is turned off by default. 
2. Add runtimeFlags2 to J9SharedClassConfig.
3. Add a new mode into J9SharedClassCacheMode.
4. Turn on -Xshareclasses automatically by -XX:+ShareOrphans
5. Add test cases to make sure that classes from SCC-unaware class loader are stored into SCC and loaded in subsequent runs via byte-to-byte comparison if this feature is turned on.

Closes #19323

### CODE REVIEW

File: test/functional/cmdLineTests/utils/src/org/openj9/test/classtests/CustomClassLoader.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import java.io.IOException;
+
+public class CustomClassLoader extends ClassLoader {
+	public CustomClassLoader() {
+		super(null);
+	}
+
+	protected Class<?> findClass(String className) throws ClassNotFoundException {
+		/*
+		 * This clas loader does not call Java SCC Helper APIs. It is not a subclass of URLClassLoader either.
```
Comment (ThanHenderson) [inline test/functional/cmdLineTests/utils/src/org/openj9/test/classtests/CustomClassLoader.java:35]: clas -> class

Comment (hangshao0) [reply test/functional/cmdLineTests/utils/src/org/openj9/test/classtests/CustomClassLoader.java]: Fixed.

**Comment by pshipton** [comment]
Pls create a doc issue.

Why does `-XX:+ShareOrphans` need to be enabled in combination with `-Xshareclasses`? Would it make more sense to automatically enable `-Xshareclasses` if `-XX:+ShareOrphans` is enabled?

File: test/functional/cmdLineTests/shareClassTests/SCCMLTests/ShareClassesCMLTests-1.xml [hunk 2]
```diff
@@ -1462,6 +1465,88 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-ex
...
+		<output type="failure" caseSensitive="no" regex="no">Unhandled Exception</output>
+		<output type="failure" caseSensitive="yes" regex="no">Exception:</output>
+		<output type="failure" caseSensitive="no" regex="no">corrupt</output>
+		<output type="failure" caseSensitive="yes" regex="no">Processing dump event</output>
+		<output type="failure" caseSensitive="yes" regex="no">JVM requested Java dump</output>
+		<output type="failure" caseSensitive="yes" regex="no">JVM requested Snap dump</output>
+	</test>
+
+	<test id="Test 69-a: Rerun the test with -XX:+ShareOrphan" timeout="600" runPath=".">
+		<command>$JAVA_EXE$ $currentMode$ -XX:+ShareOrphan $CP_UTILS$ $PROGRAM_ORPHANTEST$</command>
```
Comment (pshipton) [inline test/functional/cmdLineTests/shareClassTests/SCCMLTests/ShareClassesCMLTests-1.xml:1503]: You could add `-XX:-ShareOrphan` here before `-XX:+ShareOrphan`, which won't change the test, but give a little testing for the option.

File: test/functional/cmdLineTests/shareClassTests/SCCMLTests/ShareClassesCMLTests-1.xml [hunk 3]
```diff
@@ -1462,6 +1465,88 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-ex
...
+		<output type="failure" caseSensitive="no" regex="no">Unhandled Exception</output>
+		<output type="failure" caseSensitive="yes" regex="no">Exception:</output>
+		<output type="failure" caseSensitive="no" regex="no">corrupt</output>
+		<output type="failure" caseSensitive="yes" regex="no">Processing dump event</output>
+		<output type="failure" caseSensitive="yes" regex="no">JVM requested Java dump</output>
+		<output type="failure" caseSensitive="yes" regex="no">JVM requested Snap dump</output>
+	</test>
+
+	<test id="Test 69-d: Make sure orpahn classes from CustomClassLoader are not re-used via compareROMClassForEquality in default mode" timeout="600" runPath=".">
+		<command>$JAVA_EXE$ $currentMode$ -Xtrace:print={j9bcu.270} $CP_UTILS$ $PROGRAM_ORPHANTEST$</command>
```
Comment (pshipton) [inline test/functional/cmdLineTests/shareClassTests/SCCMLTests/ShareClassesCMLTests-1.xml:1545]: Pls add `-XX:-ShareOrphans` here, or add a second test with that option.

**Comment by hangshao0** [comment]
> Would it make more sense to automatically enable -Xshareclasses if -XX:+ShareOrphans is enabled?

I think we can let -XX:+ShareOrphans automatically turn on -Xshareclasses.

**Comment by hangshao0** [comment]
> Pls create a doc issue.

https://github.com/eclipse-openj9/openj9-docs/issues/1330

File: runtime/nls/exel/exelib.nls [hunk 4]
```diff
@@ -945,4 +945,16 @@ J9NLS_EXELIB_INTERNAL_HELP_XXNOPORTABLESHAREDCACHE.system_action=
 J9NLS_EXELIB_INTERNAL_HELP_XXNOPORTABLESHAREDCACHE.user_response=
 # END NON-TRANSLATABLE
 
+J9NLS_EXELIB_INTERNAL_HELP_XXENABLESHAREORPAH=\  -XX:+ShareOrphan   enable sharing classes with no class/module path information as orphan
```
[the remark this answers is not in the document]
Comment (hangshao0) [reply runtime/nls/exel/exelib.nls]: Fixed.

File: runtime/shared/shrclssup.c [hunk 5]
```diff
@@ -232,6 +233,22 @@ IDATA J9VMDllMain(J9JavaVM* vm, IDATA stage, void* reserved)
 					return J9VMDLLMAIN_FAILED;
 				}
 
+				/* Do not turn on class sharing for all class loaders if "cacheRetransformed" is specified,
+				 * as romclass is using SPRs pointing to its intermediateClassData which can be in the native memory rather than in the shared cache.
```
[the remark this answers is not in the document]
Comment (hangshao0) [reply runtime/shared/shrclssup.c]: Fixed.

File: runtime/shared/shrclssup.c [hunk 6]
```diff
@@ -232,6 +233,22 @@ IDATA J9VMDllMain(J9JavaVM* vm, IDATA stage, void* reserved)
...
+				 * Do not turn of this feature in the default mode. (i.e. -Xsharclasses is not specified in the command line)
```
[the remark this answers is not in the document]
Comment (hangshao0) [reply runtime/shared/shrclssup.c]: Fixed.

File: runtime/shared_common/shrinit.cpp [hunk 7]
```diff
@@ -3942,9 +3949,11 @@ j9shr_getCacheSizeBytes(J9JavaVM *vm)
 }
 
 /**
- * Determine the type of shared class cache that is enabled. Either the current default (Bootstrap Classes Only), 
- * or user defined shared cache, enabled via "-Xshareclasses" on the command line. 
- * 
+ * Determine the type of shared class cache that is enabled.
+ * Either the current default (Bootstrap Classes Only),
+ * or classes with class path information (from class loaders that are SCC enabled via Java shared classes Helper API),
+ * or classes from all class loaders (classes with no class path info shared as orphan).
```
[the remark this answers is not in the document]
Comment (hangshao0) [reply runtime/shared_common/shrinit.cpp]: Fixed.

File: test/functional/cmdLineTests/shareClassTests/SCCMLTests/ShareClassesCMLTests-1.xml (hunk 2 above)
Comment (hangshao0) [reply test/functional/cmdLineTests/shareClassTests/SCCMLTests/ShareClassesCMLTests-1.xml]: Fixed.

File: test/functional/cmdLineTests/shareClassTests/SCCMLTests/ShareClassesCMLTests-1.xml (hunk 3 above)
Comment (hangshao0) [reply test/functional/cmdLineTests/shareClassTests/SCCMLTests/ShareClassesCMLTests-1.xml:1559]: New test added.

**Comment by hangshao0** [comment]
>> Would it make more sense to automatically enable -Xshareclasses if -XX:+ShareOrphans is enabled?

> I think we can let -XX:+ShareOrphans automatically turn on -Xshareclasses.

This change is added in. It means if we turn on -XX:+ShareOrphans by default in the future, -Xshareclasss will also be turned on.

Rebased to resolve merge conflicts. This is ready to be reviewed again.

File: runtime/shared/shrclssup.c [hunk 8]
```diff
@@ -324,7 +346,31 @@ IDATA J9VMDllMain(J9JavaVM* vm, IDATA stage, void* reserved)
...
 				return J9VMDLLMAIN_FAILED;
 			}
+		} else if (shareOrphansArgIndex2 < shareOrphansArgIndex1) {
+			/* -Xshareclasses is not specified, but -XX:+ShareOrphans is specified. */
+			vm->sharedCacheAPI->sharedCacheEnabled = TRUE;
+			/* -XX:+ShareOrphans automatically turns on -Xshareclasses. Set xShareClassesPresent to TRUE */
+			vm->sharedCacheAPI->xShareClassesPresent = TRUE;
+			parseXXOptions(vm, &runtimeFlags);
+			runtimeFlags2 |= J9SHR_RUNTIMEFLAG2_ENABLE_CACHEORPHAN;
+			if (runtimeFlags & J9SHR_RUNTIMEFLAG_ENABLE_PERSISTENT_CACHE) {
```
Comment (pshipton) [inline runtime/shared/shrclssup.c:356]: Pls add a helper function for setting the cache type, and another for the J9VM_OPT_JITSERVER section so don't have duplicated code and somebody could miss updating in the future.

File: runtime/vm/j9vm.tdf [hunk 9]
```diff
@@ -971,3 +971,5 @@ TraceEntry=Trc_VM_ComputeNPEMsgAtPC_Entry Overhead=1 Level=3 Template="ComputeNP
 TraceEvent=Trc_VM_ComputeNPEMsgAtPC_start Overhead=1 Level=3 Template="ComputeNPEMsgAtPC with bcCurrent : romClass (%p) romMethod (%p) temps (%p) bytecodeOffset (%p) bcCurrent (0x%x) npePC (0x%zx) npeFlag (%d) isMethodFlag (%d) npeMsg (%s)"
 TraceEvent=Trc_VM_ComputeNPEMsgAtPC_end Overhead=1 Level=3 Template="ComputeNPEMsgAtPC with bcCurrent : bcCurrent (0x%x) npePC (%u) npeFlag (%d) isMethodFlag (%d) npeMsg (%s)"
 TraceExit=Trc_VM_ComputeNPEMsgAtPC_Exit Overhead=1 Level=3 Template="ComputeNPEMsgAtPC : npePC (0x%zx) npeFlag (%d) isMethodFlag (%d) npeMsg (%s)"
+
+TraceEvent=Trc_VM_allocateClassLoader_SetFlag NoEnv Overhead=1 Level=1 Template="allocateClassLoader set flag J9CLASSLOADER_SHARED_CLASSES_ENABLED for class loader (%p)"
```
Comment (pshipton) [inline runtime/vm/j9vm.tdf:975]: Will we benefit from this being level 1? 3 or higher is preferred in production.

File: runtime/vm/jvminit.c [hunk 10]
```diff
@@ -2064,9 +2063,19 @@ VMInitStages(J9JavaVM *vm, IDATA stage, void* reserved)
 				argIndexXXNoJITServerAOTCacheIgnoreLocalSCC = FIND_ARG_IN_VMARGS(EXACT_MATCH, "-XX:-JITServerAOTCacheIgnoreLocalSCC", NULL);
 #endif /* defined(J9VM_OPT_JITSERVER) */
 
-				if (((!J9_SHARED_CACHE_DEFAULT_BOOT_SHARING(vm)) && (argIndex < 0))
-					|| (TRUE == sharedClassDisabled)
+				if ((!sharedClassDisabled)
```
Comment (pshipton) [inline runtime/vm/jvminit.c:2066]: I don't think the extra brackets are required here.

File: runtime/vm/jvminit.c [hunk 11]
```diff
@@ -2064,9 +2063,19 @@ VMInitStages(J9JavaVM *vm, IDATA stage, void* reserved)
...
+					&&(!J9_SHARED_CACHE_DEFAULT_BOOT_SHARING(vm))
```
Comment (pshipton) [inline runtime/vm/jvminit.c:2067]: Space after &&

File: runtime/shared/shrclssup.c (hunk 8 above)
Comment (hangshao0) [reply runtime/shared/shrclssup.c]: Added a new function.

File: runtime/vm/j9vm.tdf (hunk 9 above)
Comment (hangshao0) [reply runtime/vm/j9vm.tdf]: Changed to level 3.

**Review by pshipton — APPROVED** [review]

**Pull request merged**